sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -467,10 +494,148 @@ The `node` class provides the {crs}[common reference semantics].
+[cols="2a,a"]
+|===
+|Member Function|Description
+
+|
+[source,c++]
+----
+template <int Dimensions>
+void update_nd_range(nd_range<Dimensions> executionRange);
+----
+| Updates the ND-Range for this node with a new value. This new value will not
+affect any executable graphs this node is part of until it is passed to the
+executable graph's update function.
+See <<executable-graph-update, Executable Graph Update>> for more information
+about updating node parameters.
+
+Parameters:
+
+* `executionRange` - The new value for the ND-Range.
+
+Exceptions:
+
+* Throws with error code `nd_range` if `executionRange` is an invalid ND-Range
+  for this node.
+
+* Throws with error code `invalid` if the type of the node is not a kernel
+  execution.
+
+|===
+
+==== Dynamic Parameters
+
+Dynamic parameters are arguments to a node's command-group which can be updated
+by the user after the node has been added to a graph. Updating the value of a
+dynamic parameter will be reflected in the modifiable graph which contains this
+node. These updated nodes can then be passed to an executable graph to update
+it with new values.
+
+The type of the underlying object a dynamic parameter represents is set at
+compile time using a template parameter. This underlying type can be an
+accessor, a pointer to a USM allocation, scalar passed by value, or a raw byte
+representation of the argument. The raw byte representation is intended to
+enable updating arguments set using
+link:../proposed/sycl_ext_oneapi_raw_kernel_arg.asciidoc[sycl_ext_oneapi_raw_kernel_arg].
+
+Dynamic parameters are registered with nodes in a modifiable graph, with each
+registration associating one or more node arguments to the dynamic parameter
+instance. Registration happens inside the command-group that the node represents
+, and is done using the argument index (known when argument originally set using
+`set_arg()`/`set_args()`). It is valid for a node argument to be registered with
+more than one dynamic parameter instance.
+
+Registration is done by using the argument index (known when argument
+originally set using `set_arg()`/`set_args()`).
+
+See <<executable-graph-update, Executable Graph Update>> for more information
+about updating node parameters.
+
+The `dynamic_parameter` class provides the {crs}[common reference semantics].
+
+Table {counter: tableNumber}. Member functions of the `dynamic_parameter` class.
+[cols="2a,a"]
+|===
+|Member Function|Description
+
+|
+[source,c++]
+----
+dynamic_parameter(command_graph<graph_state::modifiable> graph);
+----
+|Constructs a dynamic parameter object that can be registered with node
+arguments.
+
+Parameters:
+
+* `graph` - Graph containing the nodes which will have dynamic parameters.
+
+|
+[source,c++]
+----
+dynamic_parameter(const ValueT &initialValue, command_graph<graph_state::modifiable> graph);
+----
+|Constructs a dynamic parameter object that can be registered with node
+arguments, while providing an intiial value.
```

Comment:
The initial value constructor is indeed just for deducing the template type for `dynamic_parameter` and the intended usage would be as in your second example. I agree the spec should clarify that it is used to deduce the template parameter.

I covered the rationale for templating the `dynamic_parameter` class on the arg type in a separate comment: https://github.com/intel/llvm/pull/12486#issuecomment-1952832343

Reply:
Combing my response here with your other comment:

> The main advantage I see of having the `dynamic_parameter` class templated on the argument type is that it enables us to constrain and check the type when updating.

I agree with this, but only partially.  If the user deduces the template parameter via CTAD, then I agree that you get some type safety.  However, if the application specifies the template parameter manually as in:

```
dynamic_parameter<int*> dynParamInput(myGraph);
```

Then, there's not really any type safety because the type is still specified in two places.  If the two types don't match, it's an error.

If we really want `dynamic_parameter` to be type-safe, can we remove the non-CTAD constructor, leaving only the one with `initialValue`?  That would guide people towards writing type-safe code:

```
int myParam = /*...*/;
dynamic_parameter dynamicMyParam{graph, myParam};   // Deduces type "int"

// ...

dynamicMyParam.update(42);  // Compiler enforces updated value to be "int"
```

Of course, this causes a problem for people who want to update a kernel argument via `raw_kernel_arg`.  I think you could solve this by adding a special overload of `update`:

```
template <typename ValueT>
class dynamic_parameter {
public:
  dynamic_parameter(command_graph<graph_state::modifiable> graph, const ValueT &param);

  void register(int argIndex, const handler& cgh);

  void update(const ValueT& newValue);
  void update(const raw_kernel_arg &newValue);
};
```

Reply:
> If we really want `dynamic_parameter` to be type-safe, can we remove the non-CTAD constructor, leaving only the one with `initialValue`? That would guide people towards writing type-safe code:

I think that's a fair change and increases the type safety as you say.

> Of course, this causes a problem for people who want to update a kernel argument via `raw_kernel_arg`. I think you could solve this by adding a special overload of `update`:
> 
> ```
> template <typename ValueT>
> class dynamic_parameter {
> public:
>   dynamic_parameter(command_graph<graph_state::modifiable> graph, const ValueT &param);
> 
>   void register(int argIndex, const handler& cgh);
> 
>   void update(const ValueT& newValue);
>   void update(const raw_kernel_arg &newValue);
> };
> ```

Could the user not be required to do:

```c++
raw_kernel_arg RawArg { &SomeValue, sizeof(SomeValue)};
dynamic_parameter DynamicRawParam{Graph, RawArg};
```
It seems like if we allow updating via the overload you suggested there that we're allowing the option of bypassing the update type-safety for any `ValueT`. For example:

```c++
int IntValue = 42;
dynamic_parameter IntParam{Graph, IntValue};
float FloatValue = 77.f;

// Assuming sizeof(int) == sizeof(float)
IntParam.update(raw_kernel_arg{&FloatValue, sizeof(FloatValue)};
```
If `ValueT` has always been `raw_kernel_arg` then at least we're making it clear from the start that type safety is somewhat out the window since we're dealing with raw byte representations the whole time. Does that make sense?

Reply:
I think the use of `raw_kernel_arg` is inherently not type-safe, and we just need to live with that.  If people want to use it, we shouldn't put up artificial barriers that make it harder to use.  By having two overloads of `dynamic_parameter::update`, we open the possibility of using the type-safe overload for some updates while using the non-type-safe `raw_kernel_arg` overload for other updates.

TBH, I think some users who want to use the `raw_kernel_arg` overload will see the entire `dynamic_parameter` infrastructure as unnecessary verbosity.  This doesn't really buy them any extra type-safety if they end up using the `raw_kernel_arg` overload.  These users would probably be happier with a member function on `node` like:

```
class node {
  void update_parameter(int argIndex, const raw_kernel_arg &newValue);
};
```

Reply:
> By having two overloads of dynamic_parameter::update, we open the possibility of using the type-safe overload for some updates while using the non-type-safe raw_kernel_arg overload for other updates.

Is that something you think users would actually want in practice? I would've thought that if they were using `raw_kernel_arg` then it would make sense to always use that exclusively, but maybe I don't understand the intended use cases for that extension enough.

Either way I'm not hugely opposed to adding the overload if it is considered useful to have that option.

> TBH, I think some users who want to use the raw_kernel_arg overload will see the entire dynamic_parameter infrastructure as unnecessary verbosity. This doesn't really buy them any extra type-safety if they end up using the raw_kernel_arg overload. These users would probably be happier with a member function on node like:

I don't disagree with this necessarily but I feel like if you present that more direct option to users it will drive them to choose it over the one that provides type-safety in many cases (even if they were not planning on using `raw_kernel_args` to begin with), since there doesn't seem to be any downside of doing so. I agree it's a little less verbose perhaps, though you do then have additional book keeping burdens if you want to update multiple nodes at once since arg indices may not necessarily be the same across kernels.

IMO enforcing a little verbosity is probably worth it to steer the general user towards type-safe use.

Reply:
You are probably right that users who want `raw_kernel_arg` will use it in all cases when the parameter is updated.  I guess I don't feel so strongly about adding a second overload for `dynamic_parameter::update`.

My suspicion is that people using `raw_kernel_arg` won't see much benefit from `dynamic_parameter`, though.  They can't really get any type-safety because `raw_kernel_arg` is fundamentally not type-safe.  Therefore, the extra verbosity of `dynamic_parameter` doesn't buy them anything unless a parameter is shared among several nodes.  I suspect these users would prefer a function like `node::update_parameter`.

I don't feel so strongly that I insist on a change, though.  If you want, you could go with what you have and see what feedback you get.

Reply:
I think the advantages outweigh the disadvantages in this case, so I will leave it as is and see what the feedback is as you suggest.